airsenal/scripts/replay_season.py

```diff
@@ -72,21 +83,63 @@ def replay_season(
                 tag_prefix=tag_prefix,
                 dbsession=session,
             )
+        gw_result = {"gameweek": gw, "predictions_tag": tag}
+
         if not transfers:
             continue
         if gw == gameweek_start and new_squad:
             print("Creating initial squad...")
-            fill_initial_squad(tag, gw_range, season, fpl_team_id)
+            squad = fill_initial_squad(tag, gw_range, season, fpl_team_id)
+            # no points hits due to unlimited transfers to initialise team
+            best_strategy = {
+                "points_hit": {str(gw): 0},
+                "free_transfers": {str(gw): 0},
+                "num_transfers": {str(gw): 0},
+                "players_in": {str(gw): []},
+                "players_out": {str(gw): []},
+            }
         else:
             print("Optimising transfers...")
-            run_optimization(
+            # find best squad and the strategy for this gameweek
+            squad, best_strategy = run_optimization(
                 gw_range,
                 tag,
                 season=season,
                 fpl_team_id=fpl_team_id,
                 num_thread=num_thread,
             )
+        gw_result["starting_11"] = []
+        gw_result["subs"] = []
+        for p in squad.players:
+            if p.is_starting:
+                gw_result["starting_11"].append(p.name)
+            else:
+                gw_result["subs"].append(p.name)
+            if p.is_captain:
+                gw_result["captain"] = p.name
+            elif p.is_vice_captain:
+                gw_result["vice_captain"] = p.name
+        # obtain information about the strategy used for gameweek
+        gw_result["free_transfers"] = best_strategy["free_transfers"][str(gw)]
+        gw_result["num_transfers"] = best_strategy["num_transfers"][str(gw)]
+        gw_result["points_hit"] = best_strategy["points_hit"][str(gw)]
+        players_in = best_strategy["players_in"][str(gw)]
+        gw_result["players_in"] = (
+            [get_player_name(p) for p in players_in] if len(players_in) > 0 else []
+        )
+        players_out = best_strategy["players_out"][str(gw)]
+        gw_result["players_out"] = (
+            [get_player_name(p) for p in players_out] if len(players_in) > 0 else []
+        )
+        # compute expected and actual points for gameweek
+        exp_points = squad.get_expected_points(gw, tag)
+        gw_result["expected_points"] = exp_points - gw_result["points_hit"]
+        actual_points = squad.get_actual_points(gw, season)
+        gw_result["actual_points"] = actual_points - gw_result["points_hit"]
```

Comment:
Now we are able to obtain the points hits from the gameweek strategy as `best_strategy` is returned from `run_optimization`. I've also made it record number of free transfers, number of transfers made, and players in and out for the gameweek we're predicting - maybe this could be interesting to look at later on...